### PR TITLE
Fix: controller: Avoid election storm due to incompatible CIB

### PR DESCRIPTION
**Pull request by nrwahl2**
The DC accepts a joining node even if its local CIB manager will later reject the joining node's CIB (for example, due to schema version incompatibility). This can cause an election storm.

`do_dc_join_finalize()` calls `cib_t:cmds:sync_from()` against the joining node, syncing its CIB across the cluster. However, it may fail to apply the diff, and the DC won't notice until we get an error code via the sync callback.

Here, if a joining node has the max generation so far, we verify on the DC side that we recognize its schema name before accepting its join request. That eliminates most of the cases in which we would sync the CIB and then reject the CIB ourselves on the DC.

If the CIB sync does fail in the finalize step, then we add the node to a table of nodes whose CIB syncs have failed, and then we trigger a new election, this time nacking that node if it sends a join request. The node remains in the failed sync table until it leaves the cluster (which should happen after it's nacked).

Closes T455

### CODE REVIEW

File: daemons/controld/controld_join_dc.c [hunk 1]
```diff
@@ -284,6 +284,47 @@ compare_int_fields(xmlNode * left, xmlNode * right, const char *field)
...
+    if (pcmk__str_eq(join_from, controld_globals.our_nodename,
+                     pcmk__str_casei)) {
+        return NULL;
+    }
+
+    if (get_schema_version(validation) < 0) {
+        return crm_strdup_printf("unknown schema version %s", validation);
+
+    } else {
+        const int flags = cib_sync_call|cib_quorum_override|cib_dryrun;
```
Comment (nrwahl2) [inline daemons/controld/controld_join_dc.c:314]: We may want to make this async to reduce bottlenecks. I'm starting with a sync version as a prototype and to see if we want to do it this way. Async may be more error-prone with regard to keeping track (in all cases) of the joining node with max generation that can successfully be synced.

**Comment by nrwahl2** [comment]
The assertion is because the same `client_obj` is used twice, but the first time it's used, `do_local_notify()` sets its request ID to 0 for some reason. Not sure why we're doing that... as usual, the old commit message (13e8e218) gives no explanation of that change's relevance.
```
Nov 28 22:27:19.749 fastvm-rhel-9-0-42 pacemaker-based     [222641] (cib_process_request@based_callbacks.c:1076)        error: Completed cib_replace operation for section 'all': Update does not conform to the configured schema (rc=-203, origin=fastvm-rhel-9-0-41/crmd/53, version=0.24.23)
Nov 28 22:27:19.750 fastvm-rhel-9-0-42 pacemaker-based     [222641] (cib_process_request@based_callbacks.c:1099)        trace: Completed pre-sync update from fastvm-rhel-9-0-41/crmd/53 with local notification
Nov 28 22:27:19.750 fastvm-rhel-9-0-42 pacemaker-based     [222641] (cib_process_request@based_callbacks.c:1151)        trace: Performing local sync notification for 8f793cc8-8676-4b71-842e-9d7b2ee7b2f9
Nov 28 22:27:19.750 fastvm-rhel-9-0-42 pacemaker-based     [222641] (do_local_notify@based_callbacks.c:378)     error: nwahl: client_id=8f793cc8-8676-4b71-842e-9d7b2ee7b2f9
Nov 28 22:27:19.750 fastvm-rhel-9-0-42 pacemaker-based     [222641] (do_local_notify@based_callbacks.c:388)     error: nwahl: client_obj=0x55f88121f280
Nov 28 22:27:19.750 fastvm-rhel-9-0-42 pacemaker-based     [222641] (do_local_notify@based_callbacks.c:393)     error: nwahl: client_obj->request_id: 83
Nov 28 22:27:19.750 fastvm-rhel-9-0-42 pacemaker-based     [222641] (do_local_notify@based_callbacks.c:398)     trace: Sending response 83 to client crmd (originator of delegated request)
Nov 28 22:27:19.750 fastvm-rhel-9-0-42 pacemaker-controld  [222646] (cib_native_perform_op_delegate@cib_native.c:438)   warning: Call failed: Update does not conform to the configured schema
Nov 28 22:27:19.750 fastvm-rhel-9-0-42 pacemaker-controld  [222646] (do_dc_join_filter_offer@controld_join_dc.c:425)    error: Rejecting join-1 request from fastvm-rhel-9-0-41 (with better CIB generation than current best from fastvm-rhel-9-0-42) due to failed CIB sync test: Update does not conform to the configured schema | ref=join_request-crmd-1669703240-4
Nov 28 22:27:19.750 fastvm-rhel-9-0-42 pacemaker-controld  [222646] (compare_version@utils.c:253)       trace: 3.17.2 > 3.17.0 (3)
Nov 28 22:27:19.750 fastvm-rhel-9-0-42 pacemaker-controld  [222646] (pcmk__update_peer_expected@membership.c:980)       info: do_dc_join_filter_offer: Node fastvm-rhel-9-0-41[1] - expected state is now banned (was down)
Nov 28 22:27:19.752 fastvm-rhel-9-0-42 pacemaker-based     [222641] (cib_process_request@based_callbacks.c:973)         trace: Processing peer cib_sync operation from fastvm-rhel-9-0-41/53 on fastvm-rhel-9-0-41 intended for primary (reply=fastvm-rhel-9-0-42)
Nov 28 22:27:19.752 fastvm-rhel-9-0-42 pacemaker-based     [222641] (parse_peer_options_v2@based_callbacks.c:783)       trace: Handling cib_sync reply sent from fastvm-rhel-9-0-41 to local clients
Nov 28 22:27:19.752 fastvm-rhel-9-0-42 pacemaker-based     [222641] (cib_process_request@based_callbacks.c:1105)        trace: Completed update as secondary
Nov 28 22:27:19.752 fastvm-rhel-9-0-42 pacemaker-based     [222641] (cib_process_request@based_callbacks.c:1151)        trace: Performing local sync notification for 8f793cc8-8676-4b71-842e-9d7b2ee7b2f9
Nov 28 22:27:19.752 fastvm-rhel-9-0-42 pacemaker-based     [222641] (do_local_notify@based_callbacks.c:378)     error: nwahl: client_id=8f793cc8-8676-4b71-842e-9d7b2ee7b2f9
Nov 28 22:27:19.753 fastvm-rhel-9-0-42 pacemaker-based     [222641] (do_local_notify@based_callbacks.c:388)     error: nwahl: client_obj=0x55f88121f280
Nov 28 22:27:19.754 fastvm-rhel-9-0-42 pacemaker-based     [222641] (fail_assert_as@utils.c:375)        error: do_local_notify: Forked child [222710] to record non-fatal assertion at based_callbacks.c:392 : client_obj->request_id
```

File: daemons/controld/controld_join_dc.c (hunk 1 above)
Comment (kgaillot) [reply daemons/controld/controld_join_dc.c:314]: It looks like the relevant checks are under XML_TAG_CIB in cib_process_replace(). It might be sufficient to functionize that and call it client-side here. However we'd need the version from the current CIB and I don't think we have it here. Not sure if it's worth tracking.

Comment (kgaillot) [reply daemons/controld/controld_join_dc.c:314]: Looking at it again, I'm not sure we need to do much. It looks like do_dc_join_filter_offer() already does most of the checks we'd want, including the CIB version/generation, so maybe schema validation is all we care about?

Comment (kgaillot) [reply daemons/controld/controld_join_dc.c:314]: It's possible the get_schema_version() is sufficient. We should be able to assume the CIB has passed digest and validation on its own node. I guess we still have to handle sync_from() failure somehow in finalize_sync_callback(). Maybe calling a new election isn't the best reaction at that point.

Have you come up with a reproducer?

Comment (nrwahl2) [reply daemons/controld/controld_join_dc.c:314]: > maybe schema validation is all we care about?

Schema validation was the basic goal, except that I thought we might be even safer by dry-running the CIB sync that we'll do for real during the finalize step. Schema validation of the joining node's CIB on the DC would at least be sufficient to prevent the DC from rejecting the CIB of an accepted node.

In extreme edge cases (probably only from someone manually altering schemas on some nodes), non-DC nodes may reject a CIB that the DC validated. But in general, since the DC should be the lowest-versioned node, that shouldn't happen.


> get_schema_version()

I think so. It really depends on how paranoid we want to be. If we can assume, like you said, that the CIB passed on the joining node and that no one has manually altered any schemas, then ensuring that the schema version is known on the DC should be sufficient.

> Have you come up with a reproducer?

Yes, reproducing with a newer `validate-with` on the joining node than is present on the DC is pretty trivial. Bumping the feature set as suggested isn't necessary -- just create a higher-level schema on the joining node and alter its CIB to validate with that schema.

To hit a CIB rejection without a newer schema version required manually editing the schema on the joining node to allow a bogus element (`<extra>`), adding such an element to the joining node's CIB, and then trying to let it join. The DC recognizes the schema version (by name) but fails to apply the CIB diff after syncing it across the cluster due to the bogus element. It's a very contrived situation.

Comment (nrwahl2) [reply daemons/controld/controld_join_dc.c:314]: The assertion probably shouldn't happen regardless -- what I'm doing here doesn't seem like it should cause a fatal error. Kind of ambivalent about creating a task for it though, since it doesn't happen (as far as we know) in the existing code.

Comment (nrwahl2) [reply daemons/controld/controld_join_dc.c:314]: > Maybe calling a new election isn't the best reaction at that point.

Best things I've come up with are:
* Do a dry-run beforehand to make the real `sync_from()` failure almost impossible (although in its current form this wouldn't apply to a failure to sync our OWN CIB across the cluster); or
* Continue to call a new election... but before we do, add the failed sync-from node to a set of nodes whose CIBs we've failed to sync. During the next join sequence, either nack the nodes in that list or simply don't consider them as candidates for `max_generation_from`. This approach would let us cover all `sync_from()` failure conditions and all source nodes, without duplicating work during the filter_offer stage. It raises other questions though, like:
    * How long do we keep the set of nodes that we failed to sync from? For the life of the controller daemon? That seems too stringent. Until we lose a DC election? Store the joining node's generation XML (small) and give it another chance if some aspect of its generation XML changes on a new join attempt?
    * Is there a chance that `sync_from()` failed due to some transient issue, and so the joining node shouldn't be excluded as a candidate in the future? Maybe we could check for a `pcmk_(rc|err)_schema_validation` return code and decide based on that.

Comment (kgaillot) [reply daemons/controld/controld_join_dc.c:314]: > > Maybe calling a new election isn't the best reaction at that point.
> 
> Best things I've come up with are:
> 
>     * Do a dry-run beforehand to make the real `sync_from()` failure almost impossible (although in its current form this wouldn't apply to a failure to sync our OWN CIB across the cluster); or

I'm not too keen on the dry run, given how much overhead it would add to the join process. It can already be a considerable time while the cluster is unable to do anything.
 
>     * Continue to call a new election... but before we do, add the failed sync-from node to a set of nodes whose CIBs we've failed to sync. During the next join sequence, either nack the nodes in that set or simply exclude them as candidates for `max_generation_from`. This approach would let us cover all `sync_from()` failure conditions and all source nodes, without duplicating work during the filter_offer stage. It raises other questions though, like:

I like the idea.

Unfortunately we can't just exclude them from max_generation, because they will reject the new choice if it's lower. We have to nack them (at least in that case -- if another node as the same generation, we could try it, even if it's likely to fail too).
       
>       * How long do we keep the set of nodes that we failed to sync from?
>         
>         * For the life of the controller daemon? That seems too stringent.
>         * Until we lose a DC election?
>         * Store the joining node's generation XML (small) and give it another chance if some aspect of its generation XML changes on a new join attempt?
>         * If it's reasonable to send the failed sync-from node a lower-generation CIB, just exclude it as a `max_generation_from` candidate and try to sync another node's CIB to that node in a later join round? Then drop it from the exclusion set once that happens

For as long as the peer remains in the cluster. Presumably once we nack it, it'll leave the cluster, and we can drop it then. If it rejoins (hopefully after human intervention), we can start fresh.

>       * Is there a chance that `sync_from()` failed due to some transient issue, and so the joining node shouldn't be excluded as a candidate in the future? Maybe we could check for a `pcmk_(rc|err)_schema_validation` return code and decide based on that.

It's definitely possible. However I think we should still nack it, as it's clearly having serious difficulties worth human investigation. That might be overkill if it's just a blip, but probably better than letting a flaky node in. (Of course, if it's the DC side that's flaky, we may wind up nack'ing all other nodes. I don't see a good way around that.)

Comment (nrwahl2) [reply daemons/controld/controld_join_dc.c:314]: > I'm not too keen on the dry run

I don't like it either, to be honest, especially since it might have to be run synchronously to work right and it exposes another possible bug (the assertion) -- on top of the extra work per se. I hadn't thought of restarting the whole join process with the node excluded.

> they will reject the new choice if it's lower.

Yeah I wasn't sure about that -- the idea of excluding from max generation hinged on that answer.

> For as long as the peer remains in the cluster. Presumably once we nack it, it'll leave the cluster, and we can drop it then. If it rejoins (hopefully after human intervention), we can start fresh.

That sounds reasonable. If someone hits this somehow with feature set < 3.17.0, we won't send the nack, but I don't think that leaves us in any worse a position than the original issue. Past that feature set, the node should shut itself down. We're sort of in harm reduction mode... a human will have to fix whatever version mismatch or CIB corruption issue is to blame, and we want to prevent the election storm in the meantime.

That does remind me of another issue that needs attention (will file a task). I found that after a node shuts itself down due to receiving a nack, we fence it. I must have done my testing for the nack PR with fencing disabled. That's not a catastrophic problem, but it could be disruptive if fencing is slow or if the fenced node has cluster services enabled at boot.

> Of course, if it's the DC side that's flaky, we may wind up nack'ing all other nodes. I don't see a good way around that.

That's sort of what I was worried about. Either a problem on the DC side, or a communication problem. Anything that might cause a sync across the cluster to fail.

A related reason I went with the dry-run in the first place -- I'm remembering as I type this -- is that if some or all of the other nodes accept the CIB and the DC doesn't, then I think we're stuck with the new CIB on the other nodes. We've already synced it. We should be able to avoid a storm via the change proposed here -- we'll try each node as max generation in succession until we run out of nodes, try to use our own CIB, and either form a single-node membership or fail.

In summary, I was hoping to find a way to do an atomic CIB sync.

Comment (kgaillot) [reply daemons/controld/controld_join_dc.c:314]: < I found that after a node shuts itself down due to receiving a nack, we fence it. 

I'd consider that correct behavior. We can't probe it, so we don't know its state.

Comment (nrwahl2) [reply daemons/controld/controld_join_dc.c:314]: > I'd consider that correct behavior. We can't probe it, so we don't know its state.

works for me :)

Comment (kgaillot) [reply daemons/controld/controld_join_dc.c:314]: > In summary, I was hoping to find a way to do an atomic CIB sync.

At least the current proposal would be an improvement.

I think for an atomic sync, we'd want to build something similar to a dry run into the sync: the sync'ing node would send its cib to all nodes, all nodes would broadcast ack or nack, and if all peers ack, each node would commit the cib. Each node would hold off processing any other cib changes until the process completed. I'm thinking it would be a new cib flag that could apply to any operation. Sounds like a minefield though, so maybe just open a task.

Comment (nrwahl2) [reply daemons/controld/controld_join_dc.c:314]: That sounds much better. I was going for a crude emulation without having to implement that haha.

Added a wishlist task: https://projects.clusterlabs.org/T609

Comment (nrwahl2) [reply daemons/controld/controld_join_dc.c:314]: > For as long as the peer remains in the cluster. Presumably once we nack it, it'll leave the cluster, and we can drop it then.

Is there an existing peer-leave callback where we can add the drop? I would guess so, but you may be able to answer before I find it.

Comment (nrwahl2) [reply daemons/controld/controld_join_dc.c:314]: I guess `peer_update_callback()` is the best bet

**Comment by nrwahl2** [comment]
The new changes haven't been tested yet beyond making sure they compile. I've reworked it to no longer do a dry-run and instead use the hash table "nack" approach we discussed starting here: https://github.com/ClusterLabs/pacemaker/pull/2964#discussion_r1036108462

File: daemons/controld/controld_join_dc.c [hunk 2]
```diff
@@ -29,6 +41,72 @@ gboolean check_join_state(enum crmd_fsa_state cur_state, const char *source);
...
+{
+    if (failed_sync_nodes == NULL) {
+        failed_sync_nodes = pcmk__strikey_table(g_free, NULL);
+    }
+
+    /* Assert if the node is already in the table. That would indicate that we
+     * failed to nack it during the filter offer step.
+     */
+    assert(g_hash_table_insert(failed_sync_nodes, g_strdup(node_name),
+                               (gpointer *) &join_id));
```
Comment (nrwahl2) [inline daemons/controld/controld_join_dc.c:89]: I'm realizing this is probably unsafe because the memory will be freed for `join_id`, and I'm wondering if I have a similar mistake in the "don't duplicate failcounts" PR even though test runs succeeded there. Probably need to allocate a new int and give ownership to the hash table.

Comment (kgaillot) [reply daemons/controld/controld_join_dc.c]: As mentioned elsewhere, you can use the GPOINTER_TO_INT() etc. macros to avoid memory allocation

**Comment by nrwahl2** [comment]
It seems to work as expected.
1. We nack the client immediately if we don't recognize its schema version by name.
2. If we recognize the schema version but the `sync_from()` fails (for example, due to a mismatch of schema content, where the client permits something the DC doesn't), then we trigger a new election and nack that client in the next join round. The client gets fenced if stonith is enabled and there's an applicable stonith device.
3. After the client gets fenced in Scenario 2, if we start cluster services there and let it try to join again, we once again accept it the first time, the `sync_from()` fails, and we nack it the second time. This indicates that we're successfully dropping the client from the `failed_sync_nodes` table when it leaves the cluster.

File: daemons/controld/controld_join_dc.c [hunk 3]
```diff
@@ -29,6 +41,81 @@ gboolean check_join_state(enum crmd_fsa_state cur_state, const char *source);
...
+{
+    if (failed_sync_nodes == NULL) {
+        failed_sync_nodes = pcmk__strikey_table(g_free, NULL);
+    }
+
+    /* Assert if the node is already in the table. That would indicate that we
+     * failed to nack it during the filter offer step.
+     */
+    assert(g_hash_table_insert(failed_sync_nodes, g_strdup(node_name),
+                               GINT_TO_POINTER(join_id)));
```
[the remark this answers is not in the document]
Comment (nrwahl2) [reply daemons/controld/controld_join_dc.c:89]: whoops yep

and works for me

**Comment by nrwahl2** [comment]
addressed both review points and then rebased on current main

**Review by kgaillot — APPROVED** [review]
